src/Foundatio/Messaging/IMessageSubscriber.cs

```diff
@@ -4,19 +4,25 @@
 
 namespace Foundatio.Messaging {
     public interface IMessageSubscriber {
-        Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, CancellationToken cancellationToken = default) where T : class;
+        Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, CancellationToken cancellationToken = default, Func<T, bool> messagefilter = null) where T : class;
```

Comment:
Maybe some discussion around this... How often do we think message filter will be used (also noticed casing on this should be `messageFilter`, or maybe we call this something different like `filter`?)? I kind of like cancellation tokens always being the last argument. Either way this will break the abstraction.

Reply:
i don't know. i would use it always ;-)

maybe we can introduce a second method in the IMessageSubscriber interface like this:

```

 public interface IMessageSubscriber {


        Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, CancellationToken cancellationToken = default) where T : class;

        Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class;

    }

 public static class MessageBusExtensions {
        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Func<T, Task> handler, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => handler(msg), cancellationToken);
        }

        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Action<T> handler, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => {
                handler(msg);
                return Task.CompletedTask;
            }, cancellationToken);
        }


        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Func<T, Task> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => handler(msg), cancellationToken, filter);
        }

        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Action<T> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => {
                handler(msg);
                return Task.CompletedTask;
            }, cancellationToken, filter);
        }


    }

```

Reply:
we could change it this way:

```
   public interface IMessageSubscriber {


        Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, CancellationToken cancellationToken = default) where T : class;

        Task SubscribeAsync<T>(Func<T, CancellationToken, Task> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class;

    }

    public static class MessageBusExtensions {
        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Func<T, Task> handler, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => handler(msg), cancellationToken);
        }

        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Action<T> handler, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => {
                handler(msg);
                return Task.CompletedTask;
            }, cancellationToken);
        }

        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Func<T, Task> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => handler(msg), filter, cancellationToken);
        }

        public static Task SubscribeAsync<T>(this IMessageSubscriber subscriber, Action<T> handler, Func<T, bool> filter, CancellationToken cancellationToken = default) where T : class {
            return subscriber.SubscribeAsync<T>((msg, token) => {
                handler(msg);
                return Task.CompletedTask;
            }, filter, cancellationToken);
        }
        
    }
```

no breaking changes. only optional parameter

Reply:
I want to change subscribe to have an options object so that we can do other things on it as well. Like give it a subscription name or whether it’s durable etc. The message filter can be one of the options.